transport/grpc/config_test.go

```diff
@@ -353,12 +375,67 @@ func TestTransportSpec(t *testing.T) {
 					dialer, ok := single.Transport().(*Dialer)
 					require.True(t, ok, "expected *Dialer, got %T", single.Transport())
 					assert.Equal(t, wantOutbound.TLS, dialer.options.creds != nil)
+					if wantOutbound.WantCustomContextDialer {
+						assert.NotNil(t, dialer.options.contextDialer, "expected custom context dialer")
+					}
 				}
 			}
 		})
 	}
 }
 
+func TestContextDialerOptionUsage(t *testing.T) {
+	type attrs map[string]interface{}
+
+	lis, err := net.Listen("tcp", "127.0.0.1:0")
+	require.NoError(t, err)
+	defer lis.Close()
+	server := grpc.NewServer()
+	defer server.Stop()
+	go func() {
+		require.NoError(t, server.Serve(lis))
+	}()
+
+	dialContextInvoked := int32(0)
+	dialer := func(ctx context.Context, addr string) (net.Conn, error) {
+		atomic.AddInt32(&dialContextInvoked, 1)
```

Comment:
* you can use `go.uber.atomic` instead of the stdlib atomic package which makes this a little cleaner and enforces safety (cannot modify the value without using the appropriate atomic read/write)
 * I'm actually not sure we even need that though. once we have validated that the connection is ready, the connection should have been made and it should be safe to check the value right? we only expect 1 connection, and it should happen by the time we validate `dialContextInvoked`.

in case where you don't expect any races, it's better to avoid atomics, since a race being detected shows some bug/misunderstanding.

Reply:
Makes sense, there aren't going to be multiple connections. Removed atomic usage :-)